### PR TITLE
Add array find UDF

### PR DESCRIPTION
**Pull request by feilong-liu**
Fix https://github.com/prestodb/presto/issues/18304.

Add an UDF find for array, which returns the first array element which matches the predicate. Returns null if no match found.

* _find_first(array(T), function(T, boolean)) -> T_
 Returns the first element which returns true for `function(T, boolean)`. Return null if no match found.

* _find_first(array(T), startIndex, function(T, boolean)) -> T_
 If startIndex>0, find match starting at the startIndex in array until the end of array. If startIndex<0, find match starting
 at the absolute(startIndex) counting from last, until the start of array. Array is 1 indexed, startIndex=0 will return exception.

Test plan - (Please fill in how you tested your changes)

Add unit tests.

```
== RELEASE NOTES ==

General Changes
* Add an UDF find for array
 Finds the first array element which matches a predicated. Returns null if no match found.
 
.. function:: find_first(array(E), function(T,boolean)) -> E

    Returns the first element of ``array`` which returns true for ``function(T,boolean)``. Returns ``NULL`` if no such element exists.

.. function:: find_first(array(E), index, function(T,boolean)) -> E

    Returns the first element of ``array`` which returns true for ``function(T,boolean)``. Returns ``NULL`` if no such element exists.
    If ``index`` > 0, the search for element starts at position ``index`` until the end of array.
    If ``index`` < 0, the search for element starts at position ``abs(index)`` counting from last, until the start of array.

        SELECT find_first(ARRAY[3, 4, 5, 6], 2, x -> x > 0); -- 4
        SELECT find_first(ARRAY[3, 4, 5, 6], -2, x -> x > 0); -- 5
        SELECT find_first(ARRAY[3, 4, 5, 6], 2, x -> x < 4); -- NULL
        SELECT find_first(ARRAY[3, 4, 5, 6], -2, x -> x > 5); -- NULL
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/BuiltInTypeAndFunctionNamespaceManager.java [hunk 1]
```diff
@@ -808,6 +810,8 @@ private List<? extends SqlFunction> getBuildInFunctions(FeaturesConfig featuresC
                 .scalar(ArrayNgramsFunction.class)
                 .scalar(ArrayAllMatchFunction.class)
                 .scalar(ArrayAnyMatchFunction.class)
+                .scalar(ArrayFindFunction.class)
+                .scalar(ArrayFindWithOffsetFunction.class)
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/metadata/BuiltInTypeAndFunctionNamespaceManager.java:814]: Hmm - you should not need two functions for this, I don't think. Just make it a variadic function and check in the constructor?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlNullable;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+import io.airlift.slice.Slice;
+
+@Description("return the first element which matches the given predicate, null if no match")
+@ScalarFunction(value = "find", deterministic = true)
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java:26]: Is "find_first" is a better name? Java stream library also has it with this name.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlNullable;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+import io.airlift.slice.Slice;
+
+@Description("return the first element which matches the given predicate, null if no match")
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java:25]: return -> Return

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import com.facebook.presto.spi.function.SqlNullable;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.Boolean.TRUE;
+import static java.lang.Math.toIntExact;
+
+@Description("return the first element which matches the given predicate, null if no match")
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java:31]: return -> Return

**Review by kaikalur — CHANGES_REQUESTED** [review]
Also please see if we need these specializations or if we can just use boxed types and write a single method (for example like we do in: ArrayReduceFunction)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    private static int checkedIndexToBlockPosition(Block block, long index)
+    {
+        int arrayLength = block.getPositionCount();
+        if (index == 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "SQL array indices start at 1");
+        }
+        if (Math.abs(index) > arrayLength) {
+            return -1; // -1 indicates that the element is out of range and "ELEMENT_AT" should return null
+        }
+        if (index > 0) {
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java:229]:
```
index = index > 0 ? index - 1 : arrayLength + index;
return toIntExact(index);
```

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayFindFunction.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        assertFunction("find(ARRAY ['abc', 'def', 'ayz'], -2, x -> substr(x, 1, 1) = 'a')", createVarcharType(3), "abc");
+        assertFunction(
+                "find(ARRAY [ARRAY ['abc', null, '123'], ARRAY ['def', null, '456']], -2, x -> x[2] IS NULL)",
+                new ArrayType(createVarcharType(3)),
+                asList("abc", null, "123"));
+    }
+
+    @Test
+    public void testEmpty()
+    {
```
Comment (pranjalssh) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayFindFunction.java:79]: Add a test where array is null. Like find(null, x -> true)

**Review by pranjalssh — COMMENTED** [review]
Some nits

File: presto-main/src/main/java/com/facebook/presto/metadata/BuiltInTypeAndFunctionNamespaceManager.java (hunk 1 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/metadata/BuiltInTypeAndFunctionNamespaceManager.java:814]: Get an error when trying to combine the two classes to be one, there is the same error in PR https://github.com/prestodb/presto/pull/15170 too from the conversation.

**Comment by feilong-liu** [comment]
> Also please see if we need these specializations or if we can just use boxed types and write a single method (for example like we do in: ArrayReduceFunction)

Updated the diff. I see that there are some conversations in PR https://github.com/prestodb/presto/pull/15170 about boxed type vs. implementing multiple functions regarding performance which ends up in implementing multiple implementations. Should we still keep the same tradeoff here?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java (hunk 3 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: Done.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java (hunk 4 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java]: Done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java (hunk 5 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindWithOffsetFunction.java]: Done

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayFindFunction.java (hunk 6 above)
Comment (feilong-liu) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayFindFunction.java:79]: Added test.

**Comment by kaikalur** [comment]
Also don't forget to add documentation in the rst file

**Comment by feilong-liu** [comment]
> > Also please see if we need these specializations or if we can just use boxed types and write a single method (for example like we do in: ArrayReduceFunction)
> 
> Updated the diff. I see that there are some conversations in PR #15170 about boxed type vs. implementing multiple functions regarding performance which ends up in implementing multiple implementations. Should we still keep the same tradeoff here?

Discussed with @kaikalur offline, we are to keep it consistent with PR #15170 and implement multiple functions.

**Comment by feilong-liu** [comment]
> Also don't forget to add documentation in the rst file

Add document to rst file.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java (hunk 2 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFindFunction.java]: Update to `find_first`.

**Review by pranjalssh — APPROVED** [review]
Dont have strong opinions of 2 vs 1 file. You could write multiple @ScalarFunction annotations but not very clean

**Comment by kaikalur** [comment]
I don't think we should call it FIND_FIRST. Just call it FIND - we have things like ARRAY_POSITION that return the first position.

**Comment by kaikalur** [comment]
> I don't think we should call it FIND_FIRST. Just call it FIND - we have things like ARRAY_POSITION that return the first position.

Actually, it's stronger than that - the implementation returns the first element satisfying the property. But that's not the intent of the function. It is to just find an element that satisfies the predicate

**Review by kaikalur — APPROVED** [review]
Looks good!

**Pull request merged**